### PR TITLE
[s] [paid code] Communal Resource Storage [DNM]

### PR DESCRIPTION
**Pull request by Iamgoofball**
:cl: Iamgoofball
fix: The R&D console no longer has a separate duplicated UI for protolathes and circuit imprinters.
tweak: You now can use circuit imprinters and protolathes directly to print stuff.
add: All materials on the station are now stored inside the communal Ore Silo. This is housed in the Vault which has been relocated to Cargo.
add: All protolathes, techfabs, and circuit imprinters now pull resources from the Ore Silo. They no longer have their own material storage. Stuffing sheets into them will put the resources into the Ore Silo.
tweak: To get sheets/alloys, print the added sheet/alloy designs that are available at the cargo protolathe.
tweak: The ORM will still give miners their points for the ores they process, but the ores will automatically be put into the Ore Silo.
add: The Quartermaster and HoP can blacklist lathes from the pool. Newly constructed non-roundstart lathes will NOT be allowed access by default. They will also be able to access logs of what was constructed and for what resources from a computer next to the Ore Silo.
/:cl:

No more material hogging, if mining has it, everyone has it.

TODO:

- [ ] GUI for blacklisting 
- [ ] Logging system/GUI for that
- [ ] More checks to ensure an Ore Silo exists before acting on it
- [ ] Add stuffing into lathes -> ore silo
- [ ] Make new lathes link with multitools

### CODE REVIEW

File: code/modules/mining/ore_storage_bin.dm [hunk 1]
```diff
@@ -0,0 +1,20 @@
+GLOBAL_DATUM(ore_silo, /obj/machinery/ore_silo)
+GLOBAL_LIST_INIT(silo_access_lathes, list())
```
Comment (Cyberboss) [inline code/modules/mining/ore_storage_bin.dm]: GLOBAL_LIST_EMTPY

File: code/modules/mining/ore_storage_bin.dm [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+	desc = "Stores the station's ore. "
+	icon = 'icons/obj/machines/mining_machines.dmi'
+	icon_state = "ore_silo"
+	density = TRUE
+	anchored = TRUE
+
+/obj/machinery/ore_silo/Initialize()
+	. = ..()
+	AddComponent(/datum/component/material_container, list(MAT_METAL, MAT_GLASS, MAT_SILVER, MAT_GOLD, MAT_DIAMOND, MAT_PLASMA, MAT_URANIUM, MAT_BANANIUM, MAT_TITANIUM, MAT_BLUESPACE),INFINITY, FALSE, list(/obj/item/stack))
+	GLOB.ore_silo = src
```
Comment (Cyberboss) [inline code/modules/mining/ore_storage_bin.dm]: Needs sanity checks + clear on Destroy()

File: code/modules/research/machinery/_production.dm [hunk 3]
```diff
@@ -19,13 +19,14 @@
 	var/screen = RESEARCH_FABRICATOR_SCREEN_MAIN
 	var/selected_category
 
-/obj/machinery/rnd/production/Initialize()
+/obj/machinery/rnd/production/Initialize(mapload)
 	. = ..()
+	if(mapload)
+		GLOB.silo_access_lathes[src] = TRUE
+	else
+		GLOB.silo_access_lathes[src] = FALSE
```
Comment (Cyberboss) [inline code/modules/research/machinery/_production.dm]: GLOB.silo_access_lathes[src] = mapload

**Review by Cyberboss — CHANGES_REQUESTED** [review]

**Comment by Iamgoofball** [comment]
Yeah I'm gonna set up multi-tool linking for new stuff soon, need to slap that on the todo list.

**Comment by iskyp** [comment]
@Iamgoofball does this machine have a circuit so you can recon it

**Comment by YakumoChen** [comment]
>Vault is in cargo

No. That's bad for map balance.

**Comment by Iamgoofball** [comment]
@YakumoChen Your feedback doesn't matter, this is a paid PR and the request was made by our head designer.

Get used to it.

**Comment by PKPenguin321** [comment]
@Iamgoofball it should still have circuits, just not be linkable, similar to how departmental lathes already work

**Comment by Iamgoofball** [comment]
dnm until https://github.com/tgstation/tgstation/pull/37904

**Comment by imsxz** [comment]
Make sure that it blacklists the specific circuit board from using ore silo mats, not just the techfab itself. Otherwise people will just rebuild if they get blacklisted.

OOS but does techwebs have logging/blacklisting for point spending? Nothing irritates me more than some random engi using all the points on clown tech.

**Comment by kevinz000** [comment]
why are you killing off ORMs? just make a separate version that goes to ore silo and make that the default.

**Comment by Repukan** [comment]
How does this interact with Lathes made on other z levels

**Comment by Wilchenx** [comment]
@KorPhaeron Ability to toggle ORM between normal and silo mode would basically fix all of this with minimal impact on gameplay, either by a interaction in ORM screen or by screwdriving the board.

**Comment by Iamgoofball** [comment]
Sorry I meant to say blocked access, to specifically prevent that from occuring

**Comment by Shadowflame909** [comment]
I'm ready to kill some greedy QMs if they cut off my supply of plasma. #scienceneedsplasma2

**Comment by kevinz000** [comment]
@Iamgoofball move the new ones to a subpath and adding it to techwebs as a silo ORM would be the high effort and better path imo, you can find/replace the paths in the main maps and keep everything else intact. old ones should still be constructable.

**Comment by Iamgoofball** [comment]
lmao @ science mains trying to keep their old system of ore hogging

**Comment by iskyp** [comment]
@Iamgoofball as a non science main i can say that keeping a version of the old ORM is a better idea and allows for more variety with mappers (and lets free golems do their thing)
but replacing the ORM used on station with nu-orm is a good idea

**Comment by Wilchenx** [comment]
@Iamgoofball calling everyone a science main doesn't help you win this argument

xoxo

**Comment by Tlaltecuhtli** [comment]
can you give cargo tech round start rifles an armor so they can protect the vault from hostile assistants?

**Comment by Tlaltecuhtli** [comment]
also ;
- is there a way to eject minerals from the lathe? or from the silo?
- call it  bluespace silo ok
- instead of white list /black list there should be a "limit" that qm can set

**Comment by Kromgar** [comment]
Is there anything to prevent assholes from stealing all the ore?

**Comment by kevinz000** [comment]
the real punishment of being lazy is having a wrecking goofball run through your second favorite department :+1:

**Comment by Shadowflame909** [comment]
You coders sure love departmental co-operation(escalation) also that black
list is so dumb. We all know it's gonna have cargo vs all departments (at
the very least science.) Is this what you wanted coders. Nations, but
everyone hunting for the qms id card.

**Comment by ExcessiveUseOfCobblestone** [comment]
I can't help if you don't understand opportunities for non-antags to ruin something = opportunities for antags to ruin something and not make it blatantly obvious it was an antag's doing.

But yes I do like conflict (escalation) in a game where some people play almost-exclusively to get the conflict-creating roles. If you have nothing to offer I don't see why a department has to spend their round only to appease you with no ability to change that because science just _has_ to get the same amount of mats as the chef.

The solution is to make each job mechanically important in some way to everyone else so that each job has bargaining power rather than having to be spoonfed goodies.

**Comment by Shadowflame909** [comment]
I didn't really see it like that. I more or less saw it as a way to
basically be chemistry and botany's version of "no mutagen, fuck off." Type
of deal. Except more massive. If an antagonist is doing it to make
departments go to war. I can't really complain.

On May 21, 2018 2:01 PM, "Grandmother of 50" <notifications@github.com>
wrote:

I can't help if you don't understand opportunities for non-antags to ruin
something = opportunities for antags to ruin something and not make it
blatantly obvious it was an antag's doing.

But yes I do like conflict (escalation) in a game where some people play
almost-exclusively to get the conflict-creating roles. If you have nothing
to offer I don't see why a department has to spend their round only to
appease you with no ability to change that because science just *has* to
get the same amount of mats as the chef.

The solution is to make each job mechanically important in some way to
everyone else so that each job has bargaining power rather than having to
be spoonfed goodies.

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/tgstation/tgstation/pull/37901#issuecomment-390781597>,
or mute the thread

**Comment by Shadowflame909** [comment]
You can't stop someones dreams. Only slow them down!

**Comment by ExcessiveUseOfCobblestone** [comment]
Botany/Chemistry relationship sucks because Botany has no bargaining power initially but gains a ton as well as takes Chem's away once they get mutagen because if you play botany you know how to make your own and heal basically everything, even death (tfw they even have the crucial reagent for parapen juice/ rezadone).

It used to be a good trade between mutagen/omnizine.

**Comment by Iamgoofball** [comment]
stop shit posting in my pr holy fuck

**Comment by Shadowflame909** [comment]
Literally no shitposts goof. Also have you ever heard of reverse psychology

On May 22, 2018 11:07 AM, "Iamgoofball" <notifications@github.com> wrote:

stop shit posting in my pr holy fuck

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/tgstation/tgstation/pull/37901#issuecomment-391086990>,
or mute the thread

**Comment by KorPhaeron** [comment]
@Iamgoofball just ignore them the PR is getting merged as soon as you finish it

**Comment by optimumtact** [comment]
@Shadowflame909 if you're going to reply by email, delete the existing email body first

**Comment by zxaber** [comment]
So what stops the QM from unlinking all lathes at round start, if there's no HoP/Cap? What is done about an assistant sneaking into engineering and blowing all the mats on a pile of lasers or something if there's no Sec? Are we just accepting that low-pop will be (even more) terrible?

Truth be told, I'm not really even sure what this is trying to fix.

**Comment by KorPhaeron** [comment]
If there are no security and no admins and no vigilantes and people grief or don't do their jobs then yes things won't function very well.

We can't have the game function flawlessly at 5-90 players with 50 some odd job slots that may or may not be filled. We can't design around the majority of critical roles being unfilled and whoever is left acting in bad faith.

**Comment by msgerbs** [comment]
So... Why can't people just walk to the ORM? I'm with zxaber I don't understand what this is fixing or improving.

**Comment by optimumtact** [comment]
@Iamgoofball proceed without cyberbosses changes, just manually link them

**Comment by Iamgoofball** [comment]
Aight, in LA this week so I'll do it next week

**Comment by ExcessiveUseOfCobblestone** [comment]
@msgerbs it's a decision between "Science/Mining has a monopoly on resources so no one else can print anything of worth" vs "dude why did you just build 50 advanced mops". We're trying the latter since we've played with the first for years.

**Comment by msgerbs** [comment]
But the ORM is in a public location. I've never had any problem, assuming miners are actually mining, getting stuff out of the ORM.

**Comment by ExcessiveUseOfCobblestone** [comment]
Some jobs don't have access and some don't go to the ORM because they have a perceived notion science/mining has a monopoly.

**Comment by MrStonedOne** [comment]
Please self-close pr or fix merge contents within 72 hours to avoid maintainer close.

**Comment by optimumtact** [comment]
you're not a maintainer nerd get out

**Comment by Iamgoofball** [comment]
Why does he have tagging access to begin with
This also isn't "stale", I'm just still whiteboarding out how I want the backend for connected machines to work

**Comment by MrStonedOne** [comment]
I never said I'd close it, I just said doing those things within 72 would avoid it getting closed by a maintainer.

**Comment by MrDoomBringer** [comment]
image

**Comment by Iamgoofball** [comment]
most of its done locally just havent committed it yet

**Comment by Iamgoofball** [comment]
i cant oversee the testmerge for this so I can't verify everything is working for the live build, if someone else wants to take over pm me on discord so I can give you what I have codewise

**Pull request closed without merging**